packages/client/lib/client/index.ts

```diff
@@ -157,21 +157,45 @@ export default class RedisClient<
         return new (RedisClient.extend(options))(options);
     }
 
-    static parseURL(url: string): RedisClientOptions {
+    static parseOptions<O extends RedisClientOptions>(options: O): O {
+        if (options?.url) {
+            const parsed = RedisClient.parseURL(options.url);
+            if (options.socket) {
+                if (options.socket.tls !== undefined && options.socket.tls !== parsed.socket.tls) {
+                    throw new TypeError(`tls socket option is set to ${options.socket.tls} which is mismatch with protocol or the URL ${options.url} passed`)
+                }
+                parsed.socket = Object.assign(options.socket, parsed.socket);
+            }
+
+            Object.assign(options, parsed);
+        }
+        return options;
+    }
+
+    static parseURL(url: string): RedisClientOptions & {
+        socket: Exclude<RedisClientOptions['socket'], undefined> & {
```

Comment:
union is a bit wordy, please guide me where you'd like to define an interface if you think it's not ok to inline that